abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1996 @@
+> `upstream` refers to the connection between proxy and backend.
+
+
+### Callbacks exposed by the Wasm module
+
+> **Note**
+> The same unique `stream_context_id` is shared by
+> a pair of `downstream` and `upstream` connections.
+
+
+#### `proxy_on_new_connection`
+
+* params:
+  - `i32 (uint32_t) stream_context_id`
+* returns:
+  - `i32 (`[`proxy_action_t`]`) action`
+
+Called when a new connection is established.
+
+Plugin must return one of the following values:
+- `CONTINUE` to allow the new connection to be established.
+- `PAUSE` to pause processing of the new connection.
+
+
+#### `proxy_on_downstream_data`
+
+* params:
+  - `i32 (uint32_t) stream_context_id`
+  - `i32 (size_t) data_size`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (`[`proxy_action_t`]`) action`
+
+Called for each data chunk received from downstream,
+even when the processing is paused.
+
+`data_size` represents total available size of the data that can be
+retrieved, and its value will increment if the processing is paused
+and data is buffered by the host and not forwarded upstream.
+
+Data (of `data_size`) can be retrieved and/or modified using
+[`proxy_get_buffer_bytes`] and/or [`proxy_set_buffer_bytes`]
+with `buffer_id` set to `DOWNSTREAM_DATA`.
+
+Paused downstream can be resumed using [`proxy_continue_stream`]
+or closed using [`proxy_close_stream`] with `stream_type` set to
+`DOWNSTREAM`.
+
+Plugin must return one of the following values:
+- `CONTINUE` to forward `DOWNSTREAM_DATA` buffer upstream.
+- `PAUSE` to pause processing.
+
+
+#### `proxy_on_downstream_connection_close`
+
+* params:
+  - `i32 (uint32_t) stream_context_id`
+  - `i32 (`[`proxy_peer_type_t`]`) peer_type`
+* returns:
+  - none
+
+Called when downstream connection is closed.
+
+
+#### `proxy_on_upstream_data`
+
+* params:
+  - `i32 (uint32_t) stream_context_id`
+  - `i32 (size_t) data_size`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (`[`proxy_action_t`]`) action`
+
+Called for each data chunk received from upstream,
+even when the processing is paused.
+
+`data_size` represents total available size of the data that can be
+retrieved, and its value will increment if the processing is paused
+and data is buffered by the host and not forwarded downstream.
+
+Data (of `data_size`) can be retrieved and/or modified using
+[`proxy_get_buffer_bytes`] and/or [`proxy_set_buffer_bytes`]
+with `buffer_id` set to `UPSTREAM_DATA`.
+
+Paused upstream can be resumed using [`proxy_continue_stream`]
+or closed using [`proxy_close_stream`] with `stream_type` set to
+`UPSTREAM`.
+
+Plugin must return one of the following values:
+- `CONTINUE` to forward `UPSTREAM_DATA` buffer downstream.
+- `PAUSE` to pause processing.
+
+
+#### `proxy_on_upstream_connection_close`
+
+* params:
+  - `i32 (uint32_t) stream_context_id`
+  - `i32 (`[`proxy_peer_type_t`]`) peer_type`
+* returns:
+  - none
+
+Called when upstream connection is closed.
+
+
+## HTTP
```

Comment:
For consistency, shouldn't this be `HTTP Steams` (as in `Common stream operations`)